src/fmureceiver.cpp

```diff
@@ -0,0 +1,307 @@
+        emit Disconnected(message);
+        return;
+    }
+
+    isPaused_ = false;
+    isRunning_ = true;
+    QtConcurrent::run(this, &FMUReceiver::ReceiveLoop);
+
+    isConnected_ = true;
+    emit Connected(currentDataType_);
+}
+
+void
+FMUReceiver::DisconnectRequested()
+{
+    if (!isConnected_)
+    {
+        return;
+    }
+
+    isThreadTerminated_ = false;
+    isRunning_ = false;
+
+    while (!isThreadTerminated_)
+    {
+        QThread::msleep(50);
+    }
+
+    isConnected_ = false;
+    emit Disconnected();
+}
+
+void
+FMUReceiver::ReceiveLoop()
+{
+    while (isRunning_)
+    {
+        bool msgReceived = false;
+
+        if (!isPaused_)
+        {
+            fmi2_boolean_t newStep = fmi2_true;
+            time_t timeC = time(NULL);
+            fmiStatus_ = fmi2_import_do_step(fmu_, (fmi2_real_t)timeC, hStep_, newStep);
+
+            osi::SensorData sd;
+            if (get_fmi_sensor_data_in(sd))
+            {
+                qDebug() << "FMU SensorData Receive data";
+                msgReceived = true;
+                emit MessageReceived(sd, currentDataType_);
+            }
+            else
+            {
+//                qDebug() << "FMU SensorData receiving error";
+            }
+        }
+
+        if (!msgReceived)
+        {
+            QThread::msleep(5);
+        }
+    }
+
+    isThreadTerminated_ = true;
+}
+
+bool
+FMUReceiver::initializeFMUWrapper()
+{
+    // TODO: config or use default values?
+    callbacks_.malloc = malloc;
+    callbacks_.calloc = calloc;
+    callbacks_.realloc = realloc;
+    callbacks_.free = free;
+    callbacks_.logger = fmuWrapperLogger;
+    if (logLevel_ == LogLevel::Warn)
+    {
+        callbacks_.log_level = jm_log_level_warning;
+    }
+    else
+    {
+        callbacks_.log_level = jm_log_level_debug;
+    }
+    callbacks_.context = 0;
+
+    context_ = fmi_import_allocate_context(&callbacks_);
+    fmi_version_enu_t version = fmi_import_get_fmi_version(context_, FMUPath_.c_str(), tmpPath_.c_str());
+
+    if (version != fmi_version_2_0_enu)
+    {
+        qDebug() << "The code only supports version 2.0";
+        return false;
+    }
+    return true;
+}
+
+bool
+FMUReceiver::importFMU()
+{
+    // TODO: config or use default values?
+    callBackFunctions_.logger = fmi2_log_forwarding;
+    callBackFunctions_.allocateMemory = calloc;
+    callBackFunctions_.freeMemory = free;
+    callBackFunctions_.componentEnvironment = fmu_;
+
+    jmStatus_ = fmi2_import_create_dllfmu(fmu_, fmi2_fmu_kind_cs, &callBackFunctions_);
+
+    if (jmStatus_ == jm_status_error)
+    {
+        return false;
+    }
+    return true;
+}
+
+bool
+FMUReceiver::initializeFMU()
+{
+    // TODO: config or use default values?
+    // TODO: alternative solution for start values & verify parameters for setup and initiliazation functions
+    // start values
+    fmi2_string_t instanceName = "Test CS model instance";
+
+    fmi2_string_t fmuLocation = "";
```

Comment:
fmuLocation should be changed to be FMI compliant (i.e. needs to point to resource directory location of unpacked FMU).